evm/vm/forks/sharding/guess_head_state_manager.py

```diff
@@ -0,0 +1,245 @@
+        # the collations in the chain
+        self.chain_collations = defaultdict(list)
+
+    # time related
+
+    def get_current_period(self):
+        return self.vmc.web3.eth.blockNumber // self.vmc.config['PERIOD_LENGTH']
+
+    def is_collator_in_period(self, period):
+        collator_address = self.vmc.get_eligible_proposer(
+            self.shard_id,
+            period,
+        )
+        return collator_address == self.my_address
+
+    def is_collator_in_lookahead_periods(self):
+        """
+        See if we are going to be a collator in the future periods
+        """
+        result = False
+        for future_periods in range(self.vmc.config['LOOKAHEAD_PERIODS']):
+            lookahead_period = self.get_current_period() + future_periods
+            result |= self.is_collator_in_period(lookahead_period)
+        return result
+
+    def is_late_collator_period(self):
+        current_period = self.get_current_period()
+        current_block_num = self.vmc.web3.eth.blockNumber
+        last_block_num_in_current_period = current_period * self.vmc.config['PERIOD_LENGTH']
+        return (
+            self.is_collator_in_period(current_period) and
+            (current_block_num + NUM_RESERVED_BLOCKS >= last_block_num_in_current_period)
+        )
+
+    # guess_head process related
+
+    async def verify_collation(self, head_collation_hash, collation_hash):
+        coroutine = fetch_and_verify_collation(
+            collation_hash,
+        )
+        task = asyncio.ensure_future(coroutine)
+        collation_validity = await asyncio.wait_for(task, None)
+        self.collation_validity_cache[collation_hash] = collation_validity
+        if not collation_validity:
+            self.chain_validity[head_collation_hash] = False
+
+    def fetch_candidate_head_hash(self):
+        head_collation_hash = None
+        try:
+            head_collation_dict = self.shard_tracker.fetch_candidate_head()
+            head_collation_hash = head_collation_dict['header'].hash
+        except NoCandidateHead:
+            self.logger.debug("No candidate head available, `guess_head` stops")
+        self.last_period_fetching_candidate_head = self.get_current_period()
+        return head_collation_hash
+
+    def set_collation_task(self, head_collation_hash, current_collation_hash, task):
+        self.unfinished_verifying_tasks.append(task)
+        self.collation_verifying_task[current_collation_hash] = task
+        self.chain_collations[head_collation_hash].append(current_collation_hash)
+
+    def get_chain_tasks(self, head_collation_hash):
+        # ignore tasks that are done
+        return [
+            self.collation_verifying_task[collation_hash]
+            for collation_hash in self.chain_collations[head_collation_hash]
+        ]
+
+    async def guess_head(self, enable_timing_up=False):
+        '''
+        Perform windback process.
+        returns None if there is no candidate head available in this period
```

Comment:
Returns the head collation hash if it meets the windback condition. Returns None If there's no qualified candidate head.

Reply:
Good suggestion! Should I add a "time's up" condition for this additionally?